apis/addons/v1alpha1/addons_types.go

```diff
@@ -313,6 +316,25 @@ type Addon struct {
 	Status AddonStatus `json:"status,omitempty"`
 }
 
+func (a *Addon) IsAvailable() bool {
+	return meta.IsStatusConditionTrue(a.Status.Conditions, Available)
+}
+
+func (a *Addon) SetUpgradePolicyStatus(val AddonUpgradePolicyValue) {
+	a.Status.UpgradePolicy = &AddonUpgradePolicyStatus{
+		ID:                 a.Spec.UpgradePolicy.ID,
+		Value:              val,
+		Version:            a.Spec.Version,
+		ObservedGeneration: a.Generation,
+	}
+}
+
+func (a *Addon) UpgradeCompleteForCurrentVersion() bool {
+	return a.Status.UpgradePolicy != nil &&
+		a.Status.UpgradePolicy.Version == a.Spec.Version &&
+		a.Status.UpgradePolicy.Value == AddonUpgradePolicyValueCompleted
```

Comment:
actually, I'm wondering... if the UpgradePolicy ID changes, which can happen, because someone deleted it and created a new one, we also want to take the new ID into consideration.

I'd add here `&& a.Status.UpgradePolicy.ID == a.Spec.UpgradePolicy.ID`

Reply:
I considered that as well and I think we would need confirmation from SD here.
I assume upgrade policies always transition from scheduled -> started -> completed/cancelled -> scheduled -> ... So if we trigger reconciliation for a new upgrade policy we would jump right to completed which I feel is not what clusters service expects.

I think when a new upgrade policy is created that will be included with the Addon CR and addon-operator should ignore it until an upgrade is scheduled and the new version is pushed. Then addon-operator would follow the expected state sequence. That's mostly conjecture though.

Reply:
we have to add it here as well:

https://github.com/openshift/addon-operator/pull/259/files#diff-0da0c0ad871b47984782773b5802b9a7166660d4245fbdc2308f9a7270709510R37-R39

so it can transition through all states

Reply:
I mean, here:
```
	if addon.Status.UpgradePolicy == nil || addon.Status.UpgradePolicy.Version != addon.Spec.Version {
		return r.reportUpgradeStarted(ctx, addon)
	}
```

Reply:
yeah, fair enough, it's a corner case and when/if that happens, we can require a new version to trigger the upgrade